sdks/java/io/kudu/src/main/java/org/apache/beam/sdk/io/kudu/KuduIO.java

```diff
@@ -0,0 +1,477 @@
+
+      abstract Read<T> build();
+    }
+
+    @VisibleForTesting
+    Coder<T> inferCoder(CoderRegistry coderRegistry) {
+      try {
+        return getCoder() != null
+            ? getCoder()
+            : coderRegistry.getCoder(TypeDescriptors.outputOf(getParseFn()));
+      } catch (CannotProvideCoderException e) {
+        throw new IllegalArgumentException(
+            "Unable to infer coder for output of parseFn ("
+                + TypeDescriptors.outputOf(getParseFn())
+                + "). Specify it explicitly using withCoder().",
+            e);
+      }
+    }
+
+    /** Reads from the Kudu cluster on the specified master addresses. */
+    public Read<T> withMasterAddresses(String masterAddresses) {
+      checkArgument(masterAddresses != null, "masterAddresses cannot be null or empty");
+      return builder().setMasterAddresses(Splitter.on(",").splitToList(masterAddresses)).build();
+    }
+
+    /** Reads from the specified table. */
+    public Read<T> withTable(String table) {
+      checkArgument(table != null, "table cannot be null");
+      return builder().setTable(table).build();
+    }
+
+    /** Provides the function to parse a row from Kudu into the typed object. */
+    public Read<T> withParseFn(SerializableFunction<RowResult, T> parseFn) {
+      checkArgument(parseFn != null, "parseFn cannot be null");
+      return builder().setParseFn(parseFn).build();
+    }
+
+    /** Filters the rows read from Kudu using the given predicates. */
+    public Read<T> withPredicates(List<KuduPredicate> predicates) {
+      checkArgument(predicates != null, "predicates cannot be null");
+      // reuse the kudu protobuf serialization mechanism
+      List<Common.ColumnPredicatePB> serializablePredicates =
+          predicates.stream().map(KuduPredicate::toPB).collect(Collectors.toList());
+      return builder().setSerializablePredicates(serializablePredicates).build();
+    }
+
+    /** Filters the columns read from the table to include only those specified. */
+    public Read<T> withProjectedColumns(List<String> projectedColumns) {
+      checkArgument(projectedColumns != null, "projectedColumns cannot be null");
+      return builder().setProjectedColumns(projectedColumns).build();
+    }
+
+    /** Reads from the table in batches of the specified size. */
+    public Read<T> withBatchSize(int batchSize) {
+      checkArgument(batchSize >= 0, "batchSize must not be negative");
+      return builder().setBatchSize(batchSize).build();
+    }
+
+    /**
+     * Instructs the read scan to resume a scan on another tablet server if the current server fails
+     * and faultTolerant is set to true.
+     */
+    public Read<T> withFaultTolerent(boolean faultTolerent) {
+      return builder().setFaultTolerent(faultTolerent).build();
+    }
+
+    /**
+     * Sets a {@link Coder} for the result of the parse function. This may be required if a coder
+     * can not be inferred automatically.
+     */
```

Comment:
FYI you can also just have a withOutputType taking in a TypeDescriptor, to handle the case where the coder is in the registry but the type has been erased.

Reply:
Thanks. I got a bit stumped on this and copied the approach of the TypedRead in `BigQueryIO`